parsec/core/gui/app.py

```diff
@@ -49,41 +49,48 @@ def _before_quit():
 
 
 async def _start_ipc_server(config, main_window, start_arg, result_queue):
-    new_instance_needed_qt = ThreadSafeQtSignal(main_window, "new_instance_needed", object)
-    foreground_needed_qt = ThreadSafeQtSignal(main_window, "foreground_needed")
-
-    async def cmd_handler(cmd):
-        if cmd["cmd"] == IPCCommand.FOREGROUND:
-            foreground_needed_qt.emit()
-        elif cmd["cmd"] == IPCCommand.NEW_INSTANCE:
-            new_instance_needed_qt.emit(cmd.get("start_arg"))
-        return {"status": "ok"}
-
-    while True:
-        try:
-            async with run_ipc_server(
-                cmd_handler, config.ipc_socket_file, win32_mutex_name=config.ipc_win32_mutex_name
-            ):
-                result_queue.put("started")
-                await trio.sleep_forever()
-
-        except IPCServerAlreadyRunning:
-            # Parsec is already started, give it our work then
+    try:
+        new_instance_needed_qt = ThreadSafeQtSignal(main_window, "new_instance_needed", object)
+        foreground_needed_qt = ThreadSafeQtSignal(main_window, "foreground_needed")
+
+        async def _cmd_handler(cmd):
+            if cmd["cmd"] == IPCCommand.FOREGROUND:
+                foreground_needed_qt.emit()
+            elif cmd["cmd"] == IPCCommand.NEW_INSTANCE:
+                new_instance_needed_qt.emit(cmd.get("start_arg"))
+            return {"status": "ok"}
+
+        while True:
             try:
+                async with run_ipc_server(
+                    _cmd_handler,
+                    config.ipc_socket_file,
+                    win32_mutex_name=config.ipc_win32_mutex_name,
+                ):
+                    result_queue.put_nowait("started")
```

Comment:
Maybe it's time to clean the queue communication? An enum would be nice here.

Reply:
:+1: